### ISSUE TITLE
I'd like some help with karma-domino-launcher

### ISSUE THREAD

**Issue by CaerusKaru** [karma-runner/karma#3251]
### Expected behaviour

We should be able to run Karma on Node without relying on all of the browser APIs so that we can create a launcher for NodeJS

### Actual behaviour

The karma.js script relies on many browser APIs and doesn't work easily with a NodeJS launcher (even if that launcher executes the karma.js script in a VM).

See example repo here: https://github.com/CaerusKaru/karma-domino-launcher

Basically unless you're using JSDOM, you can't make a launcher on Node without a browser.

**Comment by johnjbarton** [comment]
Does the rendering service run under Domino? I assume the service must have some DOM and DOM context apis? 

The client and context don't use a lot of API, but they live in html pages. And the client uses websockets (of course expecting to be the browser-side websockets).

You might have to use a customClientFile, http://karma-runner.github.io/3.0/config/configuration-file.html

**Comment by CaerusKaru** [comment]
Yeah that might be the best bet.

Some more context:
* The Angular SSR renderer uses Domino internally; that is it initializes it by itself and doesn’t depend on a pre-existing context. It also doesn’t expose any globals like document or window
* This started off as a domino launcher but really is more of a “karma-node-launcher” because it’s saying that the test execution context shouldn’t be a browser, it should be Node

**Comment by CaerusKaru** [comment]
As an aside though, there doesn't look like there's a reason this shouldn't be working in principle. Domino has all of the globals needed for execution, and the script actually makes it pretty far. It looks like there's just a missing piece in getting the context working.

To answer your original question, I was talking originally about the client `karma.js`, which is the one initially loaded in the browser and fetched by Domino. The following seems to work:

1. VM runs socket.io and sets up the globals for sockets
2. VM runs client `karma.js` and connects to socket io
3. A new script is added to the DOM which sets up the socket to upgrade to websockets
4. The VM executes this new script
5. Nothing yet

I've updated the repo with the new logic to show the progression. But I don't think it should require a custom client file to get this to work. It's probably only missing one or two elements.

**Comment by johnjbarton** [comment]
From the [domino](https://github.com/fgnass/domino) page:
> Domino doesn't execute scripts nor does it download external resources.

So I see that you are booting karma client side logic into the domino VM programmatically. 

Have you considered using jasmine directly? In this approach you would boot the VM as you have done, then inject your code-under-test, jasmine, your jasmine-runner code, and finally your test code.  your jasmine-runner code would define a reporter to store results in a global.  Then your boot code would poll the state for the result store to record the test results.  You may also be able to share a promise into the sandbox to avoid polling (fun to try but not essential).

Most of karma-runner's value-add is in controlling and communicating with the browser process. But you don't really have a browser or a process.

**Comment by CaerusKaru** [comment]
The thing about Jasmine is that if we wanted to do SSR testing, we could do it with Jasmine directly. Ideally we would use Karma to give us the independence of testing framework and watch mode, both things we couldn't get with Jasmine.
